File: src/main/java/shopping/product/domain/Product.java
package shopping.product.domain;

import javax.persistence.*;

import org.springframework.http.HttpStatus;
import shopping.common.converter.MoneyConverter;
import shopping.common.vo.ImageStoreType;
import shopping.exception.WooWaException;
import shopping.product.domain.vo.Image;
import shopping.common.domain.Money;

@Entity
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String name;
    @Embedded
    private Image image;
    @Convert(converter = MoneyConverter.class)
    private Money price;

    protected Product() {
    }

    public Product(String name, Image image, Money price) {
        validateNotNegative(price);
        this.name = name;
        this.image = image;
        this.price = price;
    }

    private void validateNotNegative(Money price) {
        if (price.isSmallerThan(new Money("0"))) {
            throw new WooWaException(
                "상품의 가격은 0원보다 작을 수 없습니다.",
                new IllegalArgumentException(),
                HttpStatus.BAD_REQUEST
            );
        }
    }

    public Product(String name, Image image, String price) {
        this(name, image, new Money(price));
    }

    public Product(String name, String imageUrl, String price) {
        this(name, new Image(ImageStoreType.NONE, imageUrl), new Money(price));
    }

    public Product(Long id, String name, String image, String price) {
        this(name, image, price);
        this.id = id;
    }

    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public Image getImage() {
        return image;
    }

    public Money getPrice() {
        return price;
    }

    public void updateValues(String name, String price, String imageUrl) {
        this.name = name;
        this.price = new Money(price);
        this.image = new Image(ImageStoreType.NONE, imageUrl);
    }
}

File: src/main/kotlin/shopping/product/Product.kt
package shopping.product

private val NAME_PATTERN = Regex("[a-zA-Z0-9가-힣ㄱ-ㅎㅏ-ㅣ ()\\[\\]+\\-&/_]*")
private val URL_PATTERN = Regex("https?://[\\w\\-._~:/?#\\[\\]@!$&'()*+,;=%]+")

class Product(
    val name: String,
    val price: Int,
    val imageUrl: String,
) {
    init {
        require(name.length <= 15) {
            "이름은 15자 이내로 입력하십시오."
        }
        require(name.matches(NAME_PATTERN)) {
            "허용되지 않는 특수 문자가 포함되어 있습니다"
        }
        require(imageUrl.matches(URL_PATTERN)) {
            "올바른 URL 형식이 아닙니다"
        }
    }
}
